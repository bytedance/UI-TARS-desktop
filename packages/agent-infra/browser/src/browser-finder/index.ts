/*
 * Copyright (c) 2025 Bytedance, Inc. and its affiliates.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Logger, defaultLogger } from '@agent-infra/logger';

import { getAnyEdgeStable } from 'edge-paths';
import { getAnyChromeStable } from './chrome-paths';
import { getAnyFirefoxStable } from './firefox-paths';

export class BrowserFinder {
  private logger: Logger;

  constructor(logger?: Logger) {
    this.logger = logger ?? defaultLogger;
  }

  public findBrowser(name?: 'chrome' | 'edge' | 'firefox'): string {
    const platform = process.platform;
    let browserPath: string;

    this.logger.info('Find browser on platform:', platform);

    if (!['darwin', 'win32', 'linux'].includes(platform)) {
      const error = new Error(`Unsupported platform: ${platform}`);
      this.logger.error(error.message);
      throw error;
    }

    if (name === 'chrome') {
      browserPath = this.findChrome();
    } else if (name === 'edge') {
      // https://learn.microsoft.com/en-us/microsoft-edge/puppeteer/
      browserPath = this.findEdge();
    } else if (name === 'firefox') {
      // https://pptr.dev/webdriver-bidi/#automate-with-chrome-and-firefox
      browserPath = this.findFirefox();
    } else {
      browserPath = this.findChrome();

      if (!browserPath) {
        browserPath = this.findEdge();
      }
      if (!browserPath) {
        browserPath = this.findFirefox();
      }
    }
    this.logger.info('browserPath:', browserPath);

    return browserPath;
  }

  private findChrome(): string {
    try {
      return getAnyChromeStable();
    } catch (e) {
      this.logger.error('Find Chrome Error', e);
      throw e;
    }
  }

  private findEdge(): string {
    try {
      return getAnyEdgeStable();
    } catch (e) {
      this.logger.error('Find Edge Error', e);
      throw e;
    }
  }

  private findFirefox(): string {
    try {
      return getAnyFirefoxStable();
    } catch (e) {
      this.logger.error('Find Firefox Error', e);
      throw e;
    }
  }
}
