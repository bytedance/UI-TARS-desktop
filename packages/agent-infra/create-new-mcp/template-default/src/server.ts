import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';

async function createServer(): Promise<McpServer> {
  const server = new McpServer({
    name: '{{name}}',
    version: process.env.VERSION || '0.0.1',
  });

  // === Tools ===
  server.registerTool(
    'test_tool',
    {
      description: 'Test tool',
      inputSchema: {
        hello: z.string().describe('Hello'),
      },
    },
    async (args) => {
      return {
        isError: false,
        content: [
          {
            type: 'text',
            text: 'Hello, ' + args.hello,
          },
        ],
      };
    },
  );

  return server;
}

export { createServer };
