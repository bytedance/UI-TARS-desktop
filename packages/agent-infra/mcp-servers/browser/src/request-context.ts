/**
 * Copyright (c) 2025 Bytedance, Inc. and its affiliates.
 * SPDX-License-Identifier: Apache-2.0
 */
import { AsyncLocalStorage } from 'async_hooks';
import type { RequestContext } from 'mcp-http-server';

const requestContext = new AsyncLocalStorage<RequestContext>();

export const setRequestContext = (context: RequestContext) => {
  requestContext.enterWith(context);
};

export const getRequestContext = (): RequestContext | undefined => {
  return requestContext.getStore();
};
