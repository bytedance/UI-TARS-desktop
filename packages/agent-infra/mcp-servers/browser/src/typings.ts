import { Logger } from '@agent-infra/logger';
import { Browser, Page } from 'puppeteer-core';

export type ToolContext = {
  page: Page;
  browser: Browser;
  logger: Logger;
};
