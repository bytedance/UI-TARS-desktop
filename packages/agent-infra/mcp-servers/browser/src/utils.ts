import type { Viewport } from 'puppeteer-core';

/**
 * Parse proxy url to username and password
 * @param proxyUrl - proxy url
 * @returns username and password
 */
export function parseProxyUrl(proxyUrl: string) {
  const result = { username: '', password: '' };

  try {
    const url = new URL(proxyUrl);
    result.username = url.username || '';
    result.password = url.password || '';
  } catch (error) {
    try {
      if (proxyUrl.includes('@')) {
        const protocolIndex = proxyUrl.indexOf('://');
        if (protocolIndex !== -1) {
          const authStartIndex = protocolIndex + 3;
          const authEndIndex = proxyUrl.indexOf('@');

          if (authEndIndex > authStartIndex) {
            const authInfo = proxyUrl.substring(authStartIndex, authEndIndex);
            const authParts = authInfo.split(':');

            if (authParts.length >= 2) {
              result.username = authParts[0];
              result.password = authParts[1];
            }
          }
        }
      }
    } catch (fallbackError) {
      console.error('parse proxy url error:', fallbackError);
    }
  }

  return result;
}

export function parseViewportSize(viewportSize: string): Viewport | undefined {
  if (!viewportSize || typeof viewportSize !== 'string') {
    return undefined;
  }

  const [width, height] = viewportSize.split(',').map(Number);
  return { width, height };
}

export function parserFactor(factor: string): [number, number] | undefined {
  if (!factor || typeof factor !== 'string') {
    return undefined;
  }

  const [widthFactor, heightFactor] = factor.split(',').map(Number);
  return [widthFactor, heightFactor];
}
