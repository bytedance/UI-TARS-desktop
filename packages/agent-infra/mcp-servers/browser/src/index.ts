#!/usr/bin/env node
/**
 * The following code is modified based on
 * https://github.com/modelcontextprotocol/servers/blob/main/src/puppeteer/index.ts
 *
 * MIT License
 * Copyright (c) 2024 Anthropic, PBC
 * https://github.com/modelcontextprotocol/servers/blob/main/LICENSE
 */
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { createServer, getBrowser } from './server.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';

declare global {
  interface Window {
    mcpHelper: {
      logs: string[];
      originalConsole: Partial<typeof console>;
    };
  }
}

async function runServer() {
  const server: McpServer = createServer({
    launchOptions: {
      headless: false,
    },
    logger: {
      info: (...args: any[]) => {
        server.server.notification({
          method: 'notifications/message',
          params: {
            level: 'warning',
            logger: 'mcp-server-browser',
            data: JSON.stringify(args),
          },
        });

        server.server.sendLoggingMessage({
          level: 'info',
          data: JSON.stringify(args),
        });
      },
      error: (...args: any[]) => {
        server.server.sendLoggingMessage({
          level: 'error',
          data: JSON.stringify(args),
        });
      },
      warn: (...args: any[]) => {
        server.server.sendLoggingMessage({
          level: 'warning',
          data: JSON.stringify(args),
        });
      },
      debug: (...args: any[]) => {
        server.server.sendLoggingMessage({
          level: 'debug',
          data: JSON.stringify(args),
        });
      },
    },
  });
  const transport = new StdioServerTransport();
  await server.connect(transport);
}

runServer().catch(console.error);

process.stdin.on('close', () => {
  const { browser } = getBrowser();
  console.error('Puppeteer MCP Server closed');
  browser?.close();
});
