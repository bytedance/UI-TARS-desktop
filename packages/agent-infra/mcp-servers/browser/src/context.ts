import { store } from './store.js';
import { ResourceContext, ToolContext } from './typings.js';
import { ensureBrowser } from './utils/browser.js';

export class Context {
  async getResourceContext(): Promise<ResourceContext> {
    return {
      logger: store.logger,
    };
  }

  async getToolContext(): Promise<ToolContext | null> {
    const { logger, globalConfig } = store;

    const initialBrowser = await ensureBrowser();
    const { browser } = initialBrowser;
    let { page } = initialBrowser;

    page.removeAllListeners('popup');
    page.on('popup', async (popup) => {
      if (popup) {
        logger.info(`popup page: ${popup.url()}`);
        await popup.bringToFront();
        page = popup;
        store.globalPage = popup;
      }
    });

    return {
      page,
      browser,
      logger,
      contextOptions: globalConfig.contextOptions || {},
    };
  }
}
