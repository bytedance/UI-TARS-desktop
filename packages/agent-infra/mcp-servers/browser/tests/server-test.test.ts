import { afterEach, beforeEach, describe, expect, test } from 'vitest';
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { InMemoryTransport } from '@modelcontextprotocol/sdk/inMemory.js';
import { createServer, toolsMap, type GlobalConfig } from '../src/server';

describe('Browser MCP Server', () => {
  let client: Client;
  let server: any;

  beforeEach(async () => {
    client = new Client(
      {
        name: 'test client',
        version: '1.0',
      },
      {
        capabilities: {
          roots: {
            listChanged: true,
          },
        },
      },
    );

    server = createServer({
      launchOptions: {
        headless: true,
      },
    } as GlobalConfig);

    const [clientTransport, serverTransport] =
      InMemoryTransport.createLinkedPair();

    await Promise.all([
      client.connect(clientTransport),
      server.connect(serverTransport),
    ]);
  });

  afterEach(async () => {
    await client.close();
  });

  describe('Server Configuration', () => {
    test('should list all available tools', async () => {
      const result = await client.listTools();
      expect(result.tools.map((tool) => tool.name)).toEqual(
        Object.keys(toolsMap),
      );
    });

    test('should initialize with custom config', async () => {
      const customServer = createServer({
        launchOptions: {
          headless: true,
          args: ['--no-sandbox'],
        },
        contextOptions: {
          userAgent: 'Custom User Agent',
        },
      } as GlobalConfig);
      expect(customServer).toBeDefined();
    });
  });

  describe('Navigation Operations', () => {
    test(
      'should navigate to URL successfully',
      {
        timeout: 20000,
      },
      async () => {
        const result = await client.callTool({
          name: 'browser_navigate',
          arguments: {
            url: 'https://www.bing.com',
          },
        });
        expect(result.isError).toBe(false);
      },
    );

    test(
      'should handle navigation to invalid URL',
      {
        timeout: 20000,
      },
      async () => {
        const result = await client.callTool({
          name: 'browser_navigate',
          arguments: {
            url: 'invalid-url',
          },
        });
        expect(result.isError).toBe(true);
      },
    );
  });

  describe('Page Interactions', () => {
    beforeEach(async () => {
      await client.callTool({
        name: 'browser_navigate',
        arguments: {
          url: 'https://www.bing.com',
        },
      });
    }, 20000);

    test('should get page content in different formats', async () => {
      const htmlResult = await client.callTool({
        name: 'browser_get_html',
        arguments: {},
      });
      expect(htmlResult.isError).toBe(false);

      const textResult = await client.callTool({
        name: 'browser_get_text',
        arguments: {},
      });
      expect(textResult.isError).toBe(false);

      const markdownResult = await client.callTool({
        name: 'browser_get_markdown',
        arguments: {},
      });
      expect(markdownResult.isError).toBe(false);
    });

    test('should handle screenshots', async () => {
      const result = await client.callTool({
        name: 'browser_screenshot',
        arguments: {
          name: 'test-screenshot',
          fullPage: true,
        },
      });
      expect(result.isError).toBe(false);
    });

    test('should handle keyboard interactions', async () => {
      const result = await client.callTool({
        name: 'browser_press_key',
        arguments: {
          key: 'Enter',
        },
      });
      expect(result.isError).toBe(false);
    });
  });

  describe('Tab Management', () => {
    test(
      'should manage multiple tabs',
      {
        timeout: 20000,
      },
      async () => {
        // Open new tab
        const newTabResult = await client.callTool({
          name: 'browser_new_tab',
          arguments: {
            url: 'https://www.bing.com',
          },
        });
        expect(newTabResult.isError).toBe(false);

        // List tabs
        const listResult = await client.callTool({
          name: 'browser_tab_list',
          arguments: {},
        });
        expect(listResult.isError).toBe(false);

        // Switch tab
        const switchResult = await client.callTool({
          name: 'browser_switch_tab',
          arguments: {
            index: 0,
          },
        });
        expect(switchResult.isError).toBe(false);

        // Close tab
        const closeResult = await client.callTool({
          name: 'browser_close_tab',
          arguments: {},
        });
        expect(closeResult.isError).toBe(false);
      },
    );
  });

  describe('Error Handling', () => {
    test('should handle invalid tool calls', async () => {
      await expect(
        client.callTool({
          name: 'non_existent_tool',
          arguments: {},
        }),
      ).rejects.toThrow();
    });

    test('should handle invalid tool arguments', async () => {
      const result = await client.callTool({
        name: 'browser_switch_tab',
        arguments: {
          index: -1,
        },
      });
      expect(result.isError).toBe(true);
    });
  });
});
