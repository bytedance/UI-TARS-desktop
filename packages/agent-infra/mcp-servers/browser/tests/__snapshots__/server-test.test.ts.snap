// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Browser MCP Server > Server Configuration > should list all available tools 1`] = `
[
  "browser_navigate",
  "browser_screenshot",
  "browser_click",
  "browser_form_input_fill",
  "browser_select",
  "browser_hover",
  "browser_evaluate",
  "browser_get_html",
  "browser_get_clickable_elements",
  "browser_get_text",
  "browser_get_markdown",
  "browser_read_links",
  "browser_scroll",
  "browser_go_back",
  "browser_go_forward",
  "browser_tab_list",
  "browser_new_tab",
  "browser_close",
  "browser_close_tab",
  "browser_switch_tab",
  "browser_press_key",
  "browser_get_download_list",
]
`;
